daemons/controld/controld_remote_ra.c

```diff
@@ -192,6 +194,66 @@ start_delay_helper(gpointer data)
     return FALSE;
 }
 
+/*!
+ * \internal
+ * \brief Clear node's state aside from shutdown locks.
+ *
+ * We always clear the resource history.  Transient attributes are conditionally
+ * cleared, depending on if we were told to do so by the scheduler and if the
+ * remote connection resource has not been up this whole time.
+ */
+static void
+purge_remote_node_attrs(int call_opt, const char *node_name)
+{
+    lrm_state_t *connection_rsc = NULL;
+    bool purge = false;
+    enum controld_section_e section = controld_section_all;
+
+    /* Does the remote connection resource have the purge attributes flag set?
+     * By default, it will.  It's up to the scheduler to unset the flag if it
+     * sees a meta-attr in the transition graph.
+     */
+    connection_rsc = lrm_state_find(node_name);
+    if (connection_rsc && connection_rsc->remote_ra_data) {
+        purge = remote_ra_purge_attrs(connection_rsc);
+        section = controld_section_lrm;
+    }
+
+    /* Check the uptime of connection_rsc.  If it hasn't been running long
+     * enough, set purge=true regardless of what we saw above.  "Long enough"
+     * means it started running earlier than the timestamp when we noticed
+     * it went away in the first place.
+     */
+    if (connection_rsc && connection_rsc->conn) {
+        remote_ra_data_t *ra_data = connection_rsc->remote_ra_data;
+        lrmd_t *lrm = connection_rsc->conn;
+        time_t uptime = lrmd__uptime(lrm);
+        time_t now = time(NULL);
+
+        /* Add 20s of fuzziness to give corosync a while to notice the
+         * remote host is gone.
+         */
+        if (ra_data && uptime < 20 + now - ra_data->when_host_lost) {
+            purge = true;
+        }
+    }
```

Comment:
I think we want else purge=true here (if connection_rsc or ->conn no longer exists)

Maybe also if ra_data==NULL to be paranoid, though that should never happen